clang/tools/libclang/CDependencies.cpp

```diff
@@ -418,6 +418,182 @@ CXErrorCode clang_experimental_DependencyScannerWorker_getFileDependencies_v5(
+  return Result;
+}
+
+void clang_experimental_DepGraph_dispose(CXDepGraph Graph) {
+  delete unwrap(Graph);
+}
+
+size_t clang_experimental_DepGraph_getNumModules(CXDepGraph Graph) {
+  TranslationUnitDeps &TUDeps = unwrap(Graph)->TUDeps;
+  return TUDeps.ModuleGraph.size();
+}
+
+CXDepGraphModule clang_experimental_DepGraph_getModule(CXDepGraph Graph,
+                                                       size_t Index) {
+  TranslationUnitDeps &TUDeps = unwrap(Graph)->TUDeps;
+  return wrap(&TUDeps.ModuleGraph[Index]);
+}
+
+void clang_experimental_DepGraphModule_dispose(CXDepGraphModule CXDepMod) {
+  // Part of CXDepGraph, we are not allocating extra memory for this.
+}
+
+CXString clang_experimental_DepGraphModule_getName(CXDepGraphModule CXDepMod) {
+  ModuleDeps &ModDeps = *unwrap(CXDepMod);
+  return cxstring::createRef(ModDeps.ID.ModuleName.c_str());
+}
+
+CXString
+clang_experimental_DepGraphModule_getContextHash(CXDepGraphModule CXDepMod) {
+  ModuleDeps &ModDeps = *unwrap(CXDepMod);
+  return cxstring::createRef(ModDeps.ID.ContextHash.c_str());
+}
+
+CXString
+clang_experimental_DepGraphModule_getModuleMapPath(CXDepGraphModule CXDepMod) {
+  ModuleDeps &ModDeps = *unwrap(CXDepMod);
+  return cxstring::createRef(ModDeps.ClangModuleMapFile.c_str());
+}
+
+CXStringSet *
+clang_experimental_DepGraphModule_getFileDeps(CXDepGraphModule CXDepMod) {
+  ModuleDeps &ModDeps = *unwrap(CXDepMod);
+  return cxstring::createSetRef(ModDeps.FileDeps);
+}
+
+CXStringSet *
+clang_experimental_DepGraphModule_getModuleDeps(CXDepGraphModule CXDepMod) {
+  ModuleDeps &ModDeps = *unwrap(CXDepMod);
+  std::vector<std::string> Modules;
+  for (const ModuleID &MID : ModDeps.ClangModuleDeps)
+    Modules.push_back(MID.ModuleName + ":" + MID.ContextHash);
+  return cxstring::createSet(Modules);
+}
+
+CXStringSet *
+clang_experimental_DepGraphModule_getBuildArguments(CXDepGraphModule CXDepMod) {
+  ModuleDeps &ModDeps = *unwrap(CXDepMod);
+  return cxstring::createSetRef(ModDeps.BuildArguments);
+}
+
+size_t clang_experimental_DepGraph_getNumTUCommands(CXDepGraph Graph) {
+  TranslationUnitDeps &TUDeps = unwrap(Graph)->TUDeps;
+  return TUDeps.Commands.size();
+}
+
+CXDepGraphTUCommand clang_experimental_DepGraph_getTUCommand(CXDepGraph Graph,
+                                                             size_t Index) {
+  TranslationUnitDeps &TUDeps = unwrap(Graph)->TUDeps;
+  return wrap(&TUDeps.Commands[Index]);
+}
+
+void clang_experimental_DepGraphTUCommand_dispose(CXDepGraphTUCommand CXCmd) {
+  // Part of CXDepGraph, we are not allocating extra memory for this.
+}
+
+CXString
+clang_experimental_DepGraphTUCommand_getExecutable(CXDepGraphTUCommand CXCmd) {
+  Command &TUCmd = *unwrap(CXCmd);
+  return cxstring::createRef(TUCmd.Executable.c_str());
+}
+
+CXStringSet *clang_experimental_DepGraphTUCommand_getBuildArguments(
+    CXDepGraphTUCommand CXCmd) {
+  Command &TUCmd = *unwrap(CXCmd);
+  return cxstring::createSetRef(TUCmd.Arguments);
```

Comment:
This seems a bit tricky to document since you still need to free the stringset, but the strings are only valid as long as the enclosing object itself, right? Same for the other reference types here.

Reply:
Yes, I'll add documentation to all the functions. I'd like to avoid extra copies, even if the client needs to be a bit more careful, since there are a _ton_ of strings.

Reply:
Would it make sense to return `const char *` for strings and some simple array type of `const char *; size_t` for string sets instead of CXString[Set]? It seems like it would be easier to understand if the lifetime is completely tied to the depgraph instead of partly the depgraph and partly the returned object.

Reply:
I'll take a look, it should simplify the client code a bit that way.

Reply:
See https://github.com/apple/llvm-project/pull/6520/commits/7504a5642d159f2b00083edabc6aa6530511d8de